mesonbuild/compilers/rust.py

```diff
@@ -76,13 +77,33 @@ def needs_static_linker(self) -> bool:
     def sanity_check(self, work_dir: str, environment: 'Environment') -> None:
         source_name = os.path.join(work_dir, 'sanity.rs')
         output_name = os.path.join(work_dir, 'rusttest')
-        with open(source_name, 'w', encoding='utf-8') as ofile:
-            ofile.write(textwrap.dedent(
-                '''fn main() {
-                }
-                '''))
+        cmdlist = copy(self.exelist)
 
-        cmdlist = self.exelist + ['-o', output_name, source_name]
+        with open(source_name, 'w', encoding='utf-8') as ofile:
+            # If machine kernel is not `none`, try to compile a dummy program.
+            # If 'none', this is likely a `no-std`(i.e. bare metal) project.
+            if self.info.kernel != 'none':
+                ofile.write(textwrap.dedent(
+                    '''fn main() {
+                    }
+                    '''))
+            else:
+                # If rustc linker is gcc, add `-nostartfiles`
+                if self.linker.id == 'ld.bfd':
```

Comment:
The comment and code here don't match, gcc could be used with lld, mold, gold, or bfd.

So either this should be `'ld.' in self.linker.id` because you want to check for a gnu-compatible linker, or Meson will put `'-C', 'linker=...'` in the command line universally. Alternatively, we can save reference to the C compiler in the Rust compiler for checking this if that's not sufficient.

Reply:
> this should be 'ld.' in self.linker.id because you want to check for a gnu-compatible linker
I was not sure about this and hope for a comment, thanks (i do not have ready to use toolchain with gold or mold linker).

BTW, it seems not possible to use default linker for rust :thinking: and Meson is always putting `-C linker=xxxx` even if `rust_ld` is not specified in cross file.

Reply:
> you want to check for a gnu-compatible linker

`-nostartfiles` is a `gcc` link option, not related to the underlying linker used. Is `ld.` is sufficient ?
Anyway, the only valid entries for `rust_ld` is either `'gcc'`  ou `c` (which is the cross gcc binary), with `'gold'` or `'ld'`, meson failed to check the rust toolchain with the following error:
`ERROR: Unable to get gcc pre-processor defines:`

Reply:
Sorry, I lost track of this!

 >BTW, it seems not possible to use default linker for rust 🤔 and Meson is always putting -C linker=xxxx even if rust_ld is not specified in cross file.

Meson will put `-C linker=<the C compiler or Linker you're using if you've enabled the C language>`, to ensure that you get the same linker for C and Rust code, because you can get really weird errors when you use GCC + gold and Rust + lld for linking, or even different versions of the same linker.